resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -525,9 +534,11 @@ private[yarn] class YarnAllocator(
               } catch {
                 case NonFatal(e) =>
                   logError(s"Failed to launch executor $executorId on container $containerId", e)
-                  // Assigned container should be released immediately to avoid unnecessary resource
-                  // occupation.
+                  // Assigned container should be released immediately
+                  // to avoid unnecessary resource occupation.
                   amClient.releaseAssignedContainer(containerId)
+              } finally {
+                numExecutorsStarting.decrementAndGet()
```

Comment:
`numExecutorsStarting` is incremented before the task is submitted and needs to be decremented after the task finishes, successfully or not. That's exactly what the current code does.